### PR TITLE
Add match targets and notifications

### PR DESCRIPTION
**Pull request by andresg4**
## TASKS
- [X] Added device model for notifications.
- [X] When user creates a target, returns match info, if there is any.
- [X] Added fcm gem to manage notifications using Google Firebase.

## NOTES
- [X] When frontend starts working on this, it will be necessary to make adjustments to fcm and Firebase configurations.

### CODE REVIEW

File: app/controllers/api/v1/targets_controller.rb [hunk 1]
```diff
@@ -12,6 +12,10 @@ def show; end
 
       def create
         @target = current_user.targets.create!(target_params)
+        @targets = Target.match_targets(@target)
+        devices = @targets.flat_map { |target| target.user.devices }.uniq
```
Comment (Alxzu) [inline app/controllers/api/v1/targets_controller.rb]: If you need just the ids I think you can do something like `@devices_ids = @targets.flat_map { |target| target.user.devices.pluck(:id) }.uniq`

File: app/models/target.rb [hunk 2]
```diff
@@ -12,9 +13,28 @@ class Target < ApplicationRecord
...
+  def self.match_targets(target)
+    targets_in_range = Target.where("user_id != #{target.user_id}")
+                             .where("topic_id = #{target.topic_id}")
+                             .includes(user: :devices)
+                             .within((target.radius + MAX_RADIUS),
+                                     origin: [target.latitude, target.longitude])
+    find_matches(targets_in_range, target)
+  end
+
+  def self.find_matches(targets_in_range, target)
```
Comment (Alxzu) [inline app/models/target.rb]: This seems to be a private method, isn't it?

File: spec/factories/users.rb [hunk 3]
```diff
@@ -17,6 +17,21 @@
     end
   end
 
+  factory :user_with_devices, class: User do
+    name                  { Faker::Name.name }
+    email                 { Faker::Internet.safe_email }
+    gender                { Faker::Number.between(0, 1) }
+    password              { 'password123' }
+    password_confirmation { 'password123' }
+    transient do
```
Comment (Alxzu) [inline spec/factories/users.rb:27]: Leave a space here

**Review by Alxzu — APPROVED** [review]
Excellent work! This seems nice & tidy! 💪

File: spec/requests/api/v1/targets/create_spec.rb [hunk 4]
```diff
@@ -39,6 +44,32 @@
...
+    end
+
+    context 'with matched target' do
+      it 'returns target information with matched target' do
+        params[:target][:latitude] = target.latitude + 0.0001
+        params[:target][:longitude] = target.longitude + 0.0001
+        params[:target][:topic_id] = target.topic_id
+        subject
+        new_target = Target.find_by_title(params[:target][:title])
+        expect(json).not_to be_empty
```
Comment (jdrosales17) [inline spec/requests/api/v1/targets/create_spec.rb]: This `expect` is unnecessary as the next ones include testing that the response is not empty

File: spec/requests/api/v1/targets/create_spec.rb [hunk 5]
```diff
@@ -39,6 +44,32 @@
       expect(json['longitude'].to_f).to eq(new_target.longitude.to_f)
       expect(json['topic_id']).to eq(new_target.topic_id)
       expect(json['user_id']).to eq(new_target.user_id)
+      expect(json['match_targets']).to be_empty
+    end
+
+    context 'with matched target' do
+      it 'returns target information with matched target' do
+        params[:target][:latitude] = target.latitude + 0.0001
```
Comment (jdrosales17) [inline spec/requests/api/v1/targets/create_spec.rb]: I would move this to a `let` variable

File: spec/requests/api/v1/targets/create_spec.rb [hunk 6]
```diff
@@ -39,6 +44,32 @@
...
+        new_target = Target.find_by_title(params[:target][:title])
+        expect(json).not_to be_empty
+        expect(json['id']).to eq(new_target.id)
+        expect(json['title']).to eq(new_target.title)
+        expect(json['radius']).to eq(new_target.radius)
+        expect(json['latitude'].to_f).to eq(new_target.latitude.to_f)
+        expect(json['longitude'].to_f).to eq(new_target.longitude.to_f)
+        expect(json['topic_id']).to eq(new_target.topic_id)
+        expect(json['user_id']).to eq(new_target.user_id)
+        expect(json['match_targets'][0]['id']).to eq(target.id)
```
Comment (jdrosales17) [inline spec/requests/api/v1/targets/create_spec.rb:71]: Maybe you can separate this on another `it` with the description `returns the matched targets`, just a suggestion though

**Review by jdrosales17 — APPROVED** [review]
Nice work!! Just some minor comments
